src/replicache.ts

```diff
@@ -1074,3 +1164,150 @@ type MakeMutator<
 type MakeMutators<T extends MutatorDefs> = {
   readonly [P in keyof T]: MakeMutator<T[P]>;
 };
+
+async function closeIgnoreError(tx: ReadTransactionImpl) {
+  try {
+    await tx.close();
+  } catch (ex) {
+    console.error('Failed to close transaction', ex);
+  }
+}
+
+function keyMatchesSubscription(
+  subscription: Subscription<JSONValue | undefined, unknown>,
+  indexName: string,
+  changedKey: string,
+) {
+  if (subscription.keys.has(changedKey)) {
+    return true;
+  }
+
+  for (const scanOpts of subscription.scans) {
+    if (scanOptionsMatchesKey(scanOpts, indexName, changedKey)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+function scanOptionsMatchesKey(
+  scanOpts: ScanOptionsRPC,
+  changeIndexName: string,
+  changedKey: string,
+): boolean {
+  const {
+    indexName,
+    prefix,
+    start_key: startKey,
+    start_exclusive: startExclusive,
+    start_secondary_key: startSecondaryKey,
+  } = scanOpts;
+
+  if (!indexName) {
+    if (changeIndexName) {
+      return false;
+    }
+
+    // No prefix and no start. Must recompute the subscription because all keys
+    // will have an effect on the subscription.
+    if (!prefix && !startKey) {
+      return true;
+    }
+
+    if (prefix && !changedKey.startsWith(prefix)) {
+      return false;
+    }
+
+    if (
+      startKey &&
+      ((startExclusive && changedKey <= startKey) || changedKey < startKey)
+    ) {
+      return false;
+    }
+
+    return true;
+  }
+
+  if (changeIndexName !== indexName) {
+    return false;
+  }
+
+  // No prefix and no start. Must recompute the subscription because all keys
+  // will have an effect on the subscription.
+  if (!prefix && !startKey && !startSecondaryKey) {
+    return true;
+  }
+
+  const [changedKeySecondary, changedKeyPrimary] = decodeIndexKey(changedKey);
+
+  if (prefix) {
+    if (!changedKeySecondary.startsWith(prefix)) {
+      return false;
+    }
+  }
+
+  if (
+    startSecondaryKey &&
+    ((startExclusive && changedKeySecondary <= startSecondaryKey) ||
+      changedKeySecondary < startSecondaryKey)
+  ) {
+    return false;
+  }
+
+  if (
+    startKey &&
+    ((startExclusive && changedKeyPrimary <= startKey) ||
```

Comment:
The fact that startAt uses the primary key for index scans is a bug. Can you mark it as such then fix it subsequently?

Also, once that is done these two cases (checking the primary index and checking the secondary index) can be combined I think by only considering the case where `indexName == changedIndexName`.

Reply:
This looks like it is working as intended.

- We first check the secondary key.
- Only if secondary key matches do we check the primary key.

One thing to keep in mind is that `startKey` here is the primary key and not the `start: {key}` passed into the JS scan function. This is the RPC version where things gets normalized/flattened.